src/dos/dos_classes.cpp

```diff
@@ -16,157 +19,113 @@
-	real_writed(segment+0x26,0x00,0xffffffff);		//Last File Table
-	real_writew(segment+0x26,0x04,100);				//File Table supports 100 files
-}
-
-void DOS_InfoBlock::SetFirstMCB(Bit16u _firstmcb) {
-	sSave(sDIB,firstMCB,_firstmcb); //c2woody
-}
-
-void DOS_InfoBlock::SetBuffers(Bit16u x,Bit16u y) {
-	sSave(sDIB,buffers_x,x);
-	sSave(sDIB,buffers_y,y);
-}
-
-void DOS_InfoBlock::SetCurDirStruct(Bit32u _curdirstruct) {
-	sSave(sDIB,curDirStructure,_curdirstruct);
-}
-
-void DOS_InfoBlock::SetFCBTable(Bit32u _fcbtable) {
-	sSave(sDIB,fcbTable,_fcbtable);
-}
-
-void DOS_InfoBlock::SetDeviceChainStart(Bit32u _devchain) {
-	sSave(sDIB,nulNextDriver,_devchain);
-}
-
-void DOS_InfoBlock::SetDiskBufferHeadPt(Bit32u _dbheadpt) {
-	sSave(sDIB,diskBufferHeadPt,_dbheadpt);
-}
-
-Bit16u DOS_InfoBlock::GetStartOfUMBChain(void) {
-	return (Bit16u)sGet(sDIB,startOfUMBChain);
-}
-
-void DOS_InfoBlock::SetStartOfUMBChain(Bit16u _umbstartseg) {
-	sSave(sDIB,startOfUMBChain,_umbstartseg);
-}
-
-Bit8u DOS_InfoBlock::GetUMBChainState(void) {
-	return (Bit8u)sGet(sDIB,chainingUMB);
-}
-
-void DOS_InfoBlock::SetUMBChainState(Bit8u _umbchaining) {
-	sSave(sDIB,chainingUMB,_umbchaining);
-}
-
-void DOS_InfoBlock::SetBlockDevices(Bit8u _count) {
-	sSave(sDIB,blockDevices,_count);
-}
-
-RealPt DOS_InfoBlock::GetPointer(void) {
-	return RealMake(seg,offsetof(sDIB,firstDPB));
-}
-
-Bit32u DOS_InfoBlock::GetDeviceChain(void) {
-	return sGet(sDIB,nulNextDriver);
+	SSET_WORD(sDIB, regCXfrom5e, uint16_t(0));
+	SSET_WORD(sDIB, countLRUcache, uint16_t(0));
+	SSET_WORD(sDIB, countLRUopens, uint16_t(0));
+	SSET_WORD(sDIB, protFCBs, uint16_t(0));
+	SSET_WORD(sDIB, specialCodeSeg, uint16_t(0));
+	SSET_BYTE(sDIB, joindedDrives, uint8_t(0));
+	SSET_BYTE(sDIB, lastdrive, uint8_t(0x01)); // increase this if you add drives to cds-chain
+	const RealPt dib_addr = RealMake(segment, offsetof(sDIB, diskBufferHeadPt));
+	SSET_DWORD(sDIB, diskInfoBuffer, dib_addr);
+	SSET_DWORD(sDIB, setverPtr, uint32_t(0));
+	SSET_WORD(sDIB, a20FixOfs, uint16_t(0));
+	SSET_WORD(sDIB, pspLastIfHMA, uint16_t(0));
+	SSET_BYTE(sDIB, blockDevices, uint8_t(0));
+	SSET_BYTE(sDIB, bootDrive, uint8_t(0));
+	SSET_BYTE(sDIB, useDwordMov, uint8_t(1));
+	const uint16_t ext_size = static_cast<uint16_t>(MEM_TotalPages() * 4 - 1024);
+	SSET_WORD(sDIB, extendedSize, ext_size);
+	SSET_WORD(sDIB, magicWord, uint16_t(0x0001)); // dos5+
+	SSET_WORD(sDIB, sharingCount, uint16_t(0));
+	SSET_WORD(sDIB, sharingDelay, uint16_t(0));
+	SSET_WORD(sDIB, ptrCONinput, uint16_t(0)); // no unread input available
+	SSET_WORD(sDIB, maxSectorLength, uint16_t(0x200));
+	SSET_WORD(sDIB, dirtyDiskBuffers, uint16_t(0));
+	SSET_DWORD(sDIB, lookaheadBufPt, uint32_t(0));
+	SSET_WORD(sDIB, lookaheadBufNumber, uint16_t(0));
+	SSET_BYTE(sDIB, bufferLocation, uint8_t(0)); // buffer in base memory,
+	                                             // no workspace
+	SSET_DWORD(sDIB, workspaceBuffer, uint32_t(0));
+	SSET_WORD(sDIB, minMemForExec, uint16_t(0));
```

Comment:
Looking at how almost all of these are fixed values, I wonder if we could keep a constexpr sDIB object pre-populated with all of these canned values:

constexpr sDIB set_location_defaults = { 
    regCXfrom5e = 0,
    countLRUcache = 0,
    countLRUopens = 0,
    ...
};
```

Then we assign it whenever `SetLocation` is called, and finish up with any dynamic values:

``` c++
void DOS_InfoBlock::SetLocation(uint16_t segment)
{
    ....
    sDIB obj = set_local_defaults;
    const RealPt dib_addr = RealMake(segment, offsetof(sDIB, diskBufferHeadPt));
    SSET_DWORD(sDIB, diskInfoBuffer, dib_addr);
    const uint16_t ext_size = static_cast<uint16_t>(MEM_TotalPages() * 4 - 1024);
    SSET_WORD(sDIB, extendedSize, ext_size);
```

Reply:
Hmm, I don't think so. These private packed structures (like `sDIB`) are never actually used - they are only "describing" memory layout, nothing more. So if we created a constexpr object to initialize… we would still need to keep the same litany of `SSET_*` macros to actually correctly fill the emulated memory.